ray-on-gke/user/kuberay-tpu-webhook/main.go

```diff
@@ -0,0 +1,287 @@
+	hostNames := make([]string, numWorkers)
+	for j := 0; j < numWorkers; j++ {
+		hostNames[j] = fmt.Sprintf("%s-%d.%s", workerGroupName, j, serviceName)
+	}
+	return strings.Join(hostNames, ",")
+}
+
+func injectHostnames(hostNames string, workerGroupSpec ray.WorkerGroupSpec, workerGroupIndex int, patches *[]patch) {
+	for j := 0; j < len(workerGroupSpec.Template.Spec.Containers); j++ {
+		container := workerGroupSpec.Template.Spec.Containers[j]
+		if containerRequestingTPUs(container) {
+			hostNamesPatch := patch {
+				"op": "add",
+			}
+			path := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
+			tpuWorkerHostNames := corev1.EnvVar{
+				Name:  "TPU_WORKER_HOSTNAMES",
+				Value: hostNames,
+			}
+			if len(container.Env) == 0 {
+				hostNamesPatch["path"] = path
+				hostNamesPatch["value"] = []corev1.EnvVar{tpuWorkerHostNames}
+			} else {
+				hostNamesPatch["path"] = fmt.Sprintf("%s/-", path)
+				hostNamesPatch["value"] = tpuWorkerHostNames
+			}
+			*patches = append(*patches, hostNamesPatch)
+		}
+	}
+}
+
+// add TPU_WORKER_HOSTNAMES to containers in a ray cluster
+func mutateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissionv1.AdmissionResponse, error) {
+	raycluster, err := extractRayCluster(admissionReview)
+	var patches []patch
+
+	if err == nil {
+		for i := 0; i < len(raycluster.Spec.WorkerGroupSpecs); i++ {
+			workerGroupSpec := raycluster.Spec.WorkerGroupSpecs[i]
+			if containerRequestingTPUs(workerGroupSpec.Template.Spec.Containers...) {
+				topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+				if isTPUMultiHost(topology) {
+					joinedHostNames := genDNSHostnames(workerGroupSpec)
+					injectHostnames(joinedHostNames, workerGroupSpec, i, &patches)
+				}
+			}
+		}
+	} else {
+		klog.Fatalf("Ray Cluster extraction failed: %s", err)
+	}
+
+	patchBytes, err := json.Marshal(patches)
+	if err != nil {
+		klog.Fatalf("Error serializing patches: %s", err)
+	}
+
+ 	// Create AdmissionResponse
+	admissionResponse := &admissionv1.AdmissionResponse{
+		UID: 	 admissionReview.Request.UID,
+		Allowed: true,
+		Patch:   patchBytes,
+		PatchType: func() *admissionv1.PatchType {
+			pt := admissionv1.PatchTypeJSONPatch
+			return &pt
+		}(),
+	}
+	return admissionResponse, nil
+}
+
+// unmarshal pod from admission request
+func extractPod(admissionReview *admissionv1.AdmissionReview) (*corev1.Pod, error) {
+	if admissionReview.Request.Kind.Kind != "Pod" {
+		return nil, fmt.Errorf("Expected Pod but got %s", admissionReview.Request.Kind.Kind)
+	}
+
+	pod := corev1.Pod{}
+	if err := json.Unmarshal(admissionReview.Request.Object.Raw, &pod); err != nil {
+		return nil, err
+	}
+
+	return &pod, nil
+}
+
+// add TPU_WORKER_ID to pod environment
+func mutatePod(admissionReview *admissionv1.AdmissionReview) (*admissionv1.AdmissionResponse, error) {
+	pod, err := extractPod(admissionReview)
+	var patches []patch
+	
+	if err == nil {
```

Comment:
Same here, invert the error checking logic. If the error is not nil, return it or fail immediately, don't have all the main logic inside this "if err == nil" block.

Reply:
Done.